File: exp-10/q2-dfs/Solution.java
import java.util.*;

public class Solution {
    static List<List<Integer>> adj;
    static boolean[] visited;
    static boolean first = true;

    static void dfs(int u) {
        if (!first)
            System.out.print(" ");
        first = false;
        System.out.print(u);
        visited[u] = true;
        for (int v : adj.get(u)) {
            if (!visited[v])
                dfs(v);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        adj = new ArrayList<>();
        for (int i = 0; i <= n; i++)
            adj.add(new ArrayList<>());
        for (int i = 0; i < m; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            adj.get(u).add(v);
            adj.get(v).add(u);
        }
        int s = sc.nextInt();

        for (int i = 1; i <= n; i++)
            Collections.sort(adj.get(i));

        visited = new boolean[n + 1];
        System.out.print("DFS Traversal: ");
        dfs(s);
        System.out.println();
        sc.close();
    }
}

File: exp-10/q2-dfs/solution.c
#include <stdio.h>

#define MAX 100

int adj[MAX][MAX];
int deg[MAX];
int visited[MAX];
int first = 1;

void addEdge(int u, int v)
{
    int i = deg[u];
    while (i > 0 && adj[u][i - 1] > v)
    {
        adj[u][i] = adj[u][i - 1];
        i--;
    }
    adj[u][i] = v;
    deg[u]++;
}

void dfs(int u)
{
    if (!first)
        printf(" ");
    first = 0;
    printf("%d", u);
    visited[u] = 1;
    for (int i = 0; i < deg[u]; i++)
    {
        int v = adj[u][i];
        if (!visited[v])
            dfs(v);
    }
}

int main()
{
    int n, m, s;
    scanf("%d %d", &n, &m);
    for (int i = 0; i < m; i++)
    {
        int u, v;
        scanf("%d %d", &u, &v);
        addEdge(u, v);
        addEdge(v, u);
    }
    scanf("%d", &s);

    printf("DFS Traversal: ");
    dfs(s);
    printf("\n");
    return 0;
}